contrib/platform/src/com/sun/jna/platform/win32/Kernel32Util.java

```diff
@@ -678,4 +680,141 @@ public static final String extractVolumeGUID(String volumeGUIDPath) {
+	 */
+	public static Map<String, List<String>> getResourceNames(String path) {
+		final HMODULE target = Kernel32.INSTANCE.LoadLibraryEx(path, null, Kernel32.LOAD_LIBRARY_AS_DATAFILE);
+
+		final List<String> types = new ArrayList<String>();
+		WinBase.EnumResTypeProc ertp = new WinBase.EnumResTypeProc() {
+
+			@Override
+			public boolean invoke(HMODULE module, Pointer type, Pointer lParam) {
+				// simulate IS_INTRESOURCE
+				if (Pointer.nativeValue(type) < 65535) {
+					types.add(Pointer.nativeValue(type) + "");
+				} else {
+					types.add(type.getWideString(0));
+				}
+				return true;
+			}
+		};
+
+		Kernel32.INSTANCE.EnumResourceTypes(target, ertp, null);
+
+		final Map<String, List<String>> result = new LinkedHashMap<String, List<String>>();
+
+		for (final String typeName : types) {
+			result.put(typeName, new ArrayList<String>());
+
+			Pointer pointer = null;
+			try {
+				pointer = new Pointer(Long.parseLong(typeName));
+			} catch (Exception e) {
+				pointer = new Memory(Native.WCHAR_SIZE * (typeName.length() + 1));
+				pointer.setWideString(0, typeName);
+			}
+
+			Kernel32.INSTANCE.EnumResourceNames(target, pointer, new WinBase.EnumResNameProc() {
+
+				@Override
+				public boolean invoke(HMODULE module, Pointer t, Pointer name, Pointer lParam) {
+					// simulate IS_INTRESOURCE
+					if (Pointer.nativeValue(name) < 65535) {
+						result.get(typeName).add(Pointer.nativeValue(name) + "");
+					} else {
+						result.get(typeName).add(name.getWideString(0));
+					}
+
+					return true;
+				}
+			}, null);
+		}
+
+		return result;
+	}
+
+	/**
+	 * Returns all the executable modules for a given process ID.<br>
+	 * 
+	 * @see MSDN {@link https://msdn.microsoft.com/en-us/library/windows/desktop/ms682489(v=vs.85).aspx }
+	 * @see MSDN {@link https://msdn.microsoft.com/en-us/library/windows/desktop/ms684225(v=vs.85).aspx }
+	 * @param processID
+	 *            The process ID to get executable modules for
+	 * @return All the modules in the process.
+	 */
+	public static List<Tlhelp32.MODULEENTRY32> getModules(int processID) {
+		HANDLE snapshot = Kernel32.INSTANCE.CreateToolhelp32Snapshot(Tlhelp32.TH32CS_SNAPMODULE,
+				new DWORD(processID));
+
+		List<Tlhelp32.MODULEENTRY32.ByReference> modules = new ArrayList<Tlhelp32.MODULEENTRY32.ByReference>();
+		Tlhelp32.MODULEENTRY32.ByReference first = new Tlhelp32.MODULEENTRY32.ByReference();
+		modules.add(first);
+
+		Kernel32.INSTANCE.Module32Next(snapshot, first);
+
+		Tlhelp32.MODULEENTRY32.ByReference next = new Tlhelp32.MODULEENTRY32.ByReference();
+		while (Kernel32.INSTANCE.Module32Next(snapshot, next)) {
+			modules.add(next);
+			next = new Tlhelp32.MODULEENTRY32.ByReference();
+		}
+
+		List<Tlhelp32.MODULEENTRY32> results = new ArrayList<Tlhelp32.MODULEENTRY32>();
+		for (Tlhelp32.MODULEENTRY32.ByReference mbr : modules) {
+			results.add(mbr);
+		}
+
+		Kernel32.INSTANCE.CloseHandle(snapshot);
```

Comment:
The whole method code should be braced within a _try-finally_ block:
```java
HANDLE snapshot = Kernel32.INSTANCE.CreateToolhelp32Snapshot(...);
if (...INVALID_HANDLE...) {
   throw new Win32Exception(...Kernel32.INSTANCE.GetLastError()...);
}

try {
   ...
} finally {  // avoid handle leakage
    // see https://msdn.microsoft.com/en-us/library/windows/desktop/ms724211(v=vs.85).aspx
     if (!Kernel32.INSTANCE.CloseHandle(snapshot)) {
       /*
        * Note: should re-consider this and check if we are closing the handle
        * due to an exception. If so, this exception will suppress the original one
        * whereas we want the original exception to suppress this one. It can be
       * done - if you can't figure out a simple way how let me know and I'll show you.
        */
        throw new Win32Exception(... GetLastError()...);
     }
}
```